### PR TITLE
Homework day17 improve tests

### PR DESCRIPTION
**Pull request by lyaleko**
Look please at code

### CODE REVIEW

File: src/test/java/pages/BookingPage.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    public String calendarRange = "//span[@class='sb-date-field__icon sb-date-field__icon-btn bk-svg-wrapper calendar-restructure-sb']";
+    public String selectGuests = "//label[@id='xp__guests__toggle']";
+    public String checkPricesXPath = "//button[@data-sb-id='main']";
+    public String selectCurrencyXPath = "//button[@data-modal-aria-label  = 'Выберите валюту']";
+    public String selectLanguageXPath = "//button[@data-modal-aria-label  = 'Выберите язык']";
+    public String textTooltip = "//div[@class= 'bui-tooltip__content']";
+
+    /** ResultPage **/
+
+    public String boxTenthsResultXPath = "//div[@data-testid='property-card'][10]";
```
Comment (vpetrushka) [inline src/test/java/pages/BookingPage.java]: make it dinamic

File: src/test/java/pages/BookingPage.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    public String titleTenthsResultXPath = "//div[@data-testid='property-card'][10]//div[@data-testid='title']";
+    public String menuExtraSorting = "//a[@data-type='dropdown']";
+    public String sortingByScorePrice = "//ul[@role='menu' ]//a[@data-type='review_score_and_price']";
+}
```
Comment (vpetrushka) [inline src/test/java/pages/BookingPage.java]: make them all final

File: src/test/java/pages/DemoqaPage.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+public class DemoqaPage {
+
+    /** HomePage **/
+    public String inputSelectValue = "//div[@id = 'withOptGroup']";
+    public String inputSelectOne = "//div[@id = 'selectOne']";
+    public String selectOldStyle = "//select[@id = 'oldSelectMenu']";
+    public String inputMultiselect = "//div[@class=' css-1wa3eu0-placeholder']";
+    public String selectCars = "//select[@id = 'cars']";
+}
```
Comment (vpetrushka) [inline src/test/java/pages/DemoqaPage.java]: make them all final

File: src/test/java/pages/GooglePage.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package pages;
+
+public class GooglePage {
+
+    public String searchWordInGoogleXPath = "//input[@name = 'q']";
+    public String boxResultXPath = "//div[@class = 'g']";
+}
```
Comment (vpetrushka) [inline src/test/java/pages/GooglePage.java]: make them final

File: src/test/java/steps/BasicBookingSteps.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import org.junit.Before;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import pages.BookingPage;
+
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.concurrent.TimeUnit;
+
+public class BasicBookingSteps {
```
Comment (vpetrushka) [inline src/test/java/steps/BasicBookingSteps.java]: make class abstract

File: src/test/java/steps/BasicBookingSteps.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    protected WebDriver driver = new ChromeDriver();
+    protected BookingPage page = new BookingPage();
+
+    @Before
+    public void before(){
+        setTimeouts(30);
+        driver.manage().window().maximize();
+        driver.navigate().to("https://www.booking.com/");
```
Comment (vpetrushka) [inline src/test/java/steps/BasicBookingSteps.java]: make this line dynamic

File: src/test/java/steps/BasicBookingSteps.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        driver.manage().timeouts().implicitlyWait(time, TimeUnit.SECONDS);
+        driver.manage().timeouts().pageLoadTimeout(time, TimeUnit.SECONDS);
+        driver.manage().timeouts().setScriptTimeout(time, TimeUnit.SECONDS);
+    }
+
+    public String getDateString(int days){
+        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd");
+        Calendar cal = Calendar.getInstance();
+        cal.add(Calendar.DATE, days);
+        var date = cal.getTime();
```
Comment (vpetrushka) [inline src/test/java/steps/BasicBookingSteps.java]: don't use var

File: src/test/java/tests/DemoqaTest.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        WebElement element = driver.findElement(By.xpath(page.selectOldStyle));
+
+        Select select = new Select(element);
+        select.selectByVisibleText("Aqua");
+
+        ((JavascriptExecutor)driver).executeScript(("window.scrollBy(0,1000)"));
+
+        driver.findElement(By.xpath(page.inputMultiselect)).click();
+
+        driver.findElement(By.xpath("//div[contains(text(),'Green')]")).click();
```
Comment (vpetrushka) [inline src/test/java/tests/DemoqaTest.java:44]: move all element locators out of test

File: src/test/java/steps/BasicBookingSteps.java (hunk 7 above)
Comment (lyaleko) [reply src/test/java/steps/BasicBookingSteps.java]: done

File: src/test/java/steps/BasicBookingSteps.java (hunk 6 above)
Comment (lyaleko) [reply src/test/java/steps/BasicBookingSteps.java]: done

File: src/test/java/steps/BasicBookingSteps.java (hunk 5 above)
Comment (lyaleko) [reply src/test/java/steps/BasicBookingSteps.java]: done

File: src/test/java/pages/GooglePage.java (hunk 4 above)
Comment (lyaleko) [reply src/test/java/pages/GooglePage.java]: done

File: src/test/java/pages/DemoqaPage.java (hunk 3 above)
Comment (lyaleko) [reply src/test/java/pages/DemoqaPage.java]: done

File: src/test/java/pages/BookingPage.java (hunk 2 above)
Comment (lyaleko) [reply src/test/java/pages/BookingPage.java]: done

File: src/test/java/pages/BookingPage.java (hunk 1 above)
Comment (lyaleko) [reply src/test/java/pages/BookingPage.java]: done

File: src/test/java/tests/DemoqaTest.java (hunk 8 above)
Comment (lyaleko) [reply src/test/java/tests/DemoqaTest.java:44]: done

File: src/test/java/pages/BookingPage.java [hunk 9]
```diff
@@ -4,18 +4,19 @@
...
-    public String searchPlaceXPath = "//*[@id='ss']";
+    public String searchPlaceInput = "//*[@id='ss']";
     public String calendarRange = "//span[@class='sb-date-field__icon sb-date-field__icon-btn bk-svg-wrapper calendar-restructure-sb']";
     public String selectGuests = "//label[@id='xp__guests__toggle']";
-    public String checkPricesXPath = "//button[@data-sb-id='main']";
-    public String selectCurrencyXPath = "//button[@data-modal-aria-label  = 'Выберите валюту']";
-    public String selectLanguageXPath = "//button[@data-modal-aria-label  = 'Выберите язык']";
+    public String checkPricesButton = "//button[@data-sb-id='main']";
+    public String selectCurrency = "//button[@data-modal-aria-label  = 'Выберите валюту']";
+    public String selectLanguage = "//button[@data-modal-aria-label  = 'Выберите язык']";
```
Comment (vpetrushka) [inline src/test/java/pages/BookingPage.java]: it's not selectLanguage it's selecteLanguageButton

Comment (lyaleko) [reply src/test/java/pages/BookingPage.java]: done

File: src/test/java/pages/booking/BookingHomePage.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
+package pages.booking;
+
+import driver.BasicDriver;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import utils.DateParser;
+
+public class BookingHomePage {
+
+    WebDriver driver;
```
Comment (vpetrushka) [inline src/test/java/pages/booking/BookingHomePage.java:10]: reuse existing driver

File: src/test/java/pages/booking/BookingHomePage.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    public static final String EXTRA_PARAM_INPUT = "//label[@id='xp__guests__toggle']";
+    public static final String PLUS_MINUS = "//button[@aria-label = '%s: %s количество']";
+    public static final String CLOSE_ALL_POPUPS = "//span[contains(.,'другие варианты')]";
+    public static final String SEARCH_BUTTON = "//button[@data-sb-id='main']";
+    public static final String CURRENCY_BUTTON = "//button[@data-modal-aria-label  = 'Выберите валюту']";
+    public static final String LANGUAGE_BUTTON = "//button[@data-modal-aria-label  = 'Выберите язык']";
+    public static final String TOOLTIP_TEXT = "//div[@class= 'bui-tooltip__content']";
+    public static final String DATE_CALENDAR_SELECT = "//td[@data-date='%s']";
+
+    public BookingHomePage(WebDriver driver) {
```
Comment (vpetrushka) [inline src/test/java/pages/booking/BookingHomePage.java:25]: do you need this constructor

File: src/test/java/pages/booking/BookingHomePage.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        this.driver = driver;
+    }
+
+    public void goToBookingMain() {
+        driver.navigate().to(BOOKING_SITE);
+    }
+    public void searchCity (String cityName){
```
Comment (vpetrushka) [inline src/test/java/pages/booking/BookingHomePage.java:32]: reuse inputCity method and submitSearch

File: src/test/java/pages/booking/BookingResultPage.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+    public void jsColorCard(String color, WebElement element) {
+        ((JavascriptExecutor) driver).executeScript(String.format("arguments[0].style.backgroundColor = '%s'", color), element);
+    }
+
+    public void jsColorText(String color, WebElement element) {
+        ((JavascriptExecutor) driver).executeScript(String.format("arguments[0].style.color = '%s'", color), element);
+    }
+    public double getHotelRating(){
+        return Double.parseDouble((driver.findElement(By.xpath("//div[@data-testid = 'review-score-right-component']/div[contains(@aria-label,' ')]")).getText()));
```
Comment (vpetrushka) [inline src/test/java/pages/booking/BookingResultPage.java:66]: move locator to final String

File: src/test/java/tests/MaxPriceTest.java [hunk 14]
```diff
@@ -7,51 +7,52 @@
...
 
         new FluentWait<>(driver)                                          // turn on FluentWait
                 .withTimeout(Duration.ofSeconds(5))
                 .pollingEvery(Duration.ofMillis(5))
                 .ignoring(NoSuchElementException.class)
-                .until(ExpectedConditions.invisibilityOfElementLocated(By.cssSelector(".bui-spinner.bui-spinner--size-large")));// expect spinner, then continue
+                .until(ExpectedConditions.invisibilityOfElementLocated(By.cssSelector(resultPage.SPINNER)));// expect spinner, then continue
 
-        driver.findElement(By.xpath("//a[contains(text(), 'сначала самая низкая')]")).click();
+        driver.findElement(By.xpath(resultPage.MIN_PRICE_MENU)).click();
```
Comment (vpetrushka) [inline src/test/java/tests/MaxPriceTest.java:44]: extra lines

File: src/test/java/tests/MoscowHotelsTest.java [hunk 15]
```diff
@@ -2,29 +2,18 @@
...
+        homePage.submitSearch();
 
-        WebElement webElement = new WebDriverWait(driver, 1000).until(
-                ExpectedConditions.elementToBeClickable (By.xpath("//a[contains(text(), 'сначала самая низкая')]")));
-                webElement.click();
-
-        WebElement firstElement = driver.findElement(By.xpath("//div[@data-testid='property-card'][1]"));
-
-        Assert.assertTrue(firstElement.isDisplayed());
+        Assert.assertTrue(resultPage.getDefiniteCard(1).isDisplayed());
```
Comment (vpetrushka) [inline src/test/java/tests/MoscowHotelsTest.java:17]: add message to assert

File: src/test/java/tests/ToolTipTest.java [hunk 16]
```diff
@@ -5,26 +5,29 @@
...
 
-public class ToolTipTest extends BasicBookingSteps {
 
     @Test
     public void toolTipTest() {
 
-        WebElement element = driver.findElement(By.xpath(page.selectCurrency));
+        WebElement element = driver.findElement(By.xpath(homePage.CURRENCY_BUTTON));
 
         Actions actions = new Actions(driver);
```
Comment (vpetrushka) [inline src/test/java/tests/ToolTipTest.java:20]: move to method

File: src/test/java/tests/ToolTipTest.java [hunk 17]
```diff
@@ -5,26 +5,29 @@
...
         actions.moveToElement(element);
         actions.perform();
 
-        Assert.assertEquals("Incorrect text Tooltip Currency ", "Выберите валюту", driver.findElement(By.xpath(page.textTooltip)).getText());
+        Assert.assertEquals("Incorrect text Tooltip Currency ", "Выберите валюту", driver.findElement(By.xpath(homePage.TOOLTIP_TEXT)).getText());
```
Comment (vpetrushka) [inline src/test/java/tests/ToolTipTest.java:24]: split test in two

File: src/test/java/utils/Waiter.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
+package utils;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import static driver.BasicDriver.webDriver;
+
+public class Waiter {
```
Comment (vpetrushka) [inline src/test/java/utils/Waiter.java:8]: Waiter is someone in restaurant . Be more specific

File: src/test/java/tests/MaxPriceTest.java (hunk 14 above)
Comment (lyaleko) [reply src/test/java/tests/MaxPriceTest.java:44]: done

File: src/test/java/tests/MoscowHotelsTest.java (hunk 15 above)
Comment (lyaleko) [reply src/test/java/tests/MoscowHotelsTest.java:17]: added message

File: src/test/java/utils/Waiter.java (hunk 18 above)
Comment (lyaleko) [reply src/test/java/utils/Waiter.java:8]: ;)Changed to ExplicitWait

File: src/test/java/tests/ToolTipTest.java (hunk 17 above)
Comment (lyaleko) [reply src/test/java/tests/ToolTipTest.java:24]: done

File: src/test/java/tests/ToolTipTest.java (hunk 16 above)
Comment (lyaleko) [reply src/test/java/tests/ToolTipTest.java:20]: done

File: src/test/java/pages/booking/BookingHomePage.java (hunk 12 above)
Comment (lyaleko) [reply src/test/java/pages/booking/BookingHomePage.java:32]: I have method searchCity which includes input City + search(submit) and 2 separate methodes: method input and method search. They are used for different cases.

File: src/test/java/pages/booking/BookingResultPage.java (hunk 13 above)
Comment (lyaleko) [reply src/test/java/pages/booking/BookingResultPage.java:66]: done

**Pull request merged**